flowredux/src/commonMain/kotlin/com/freeletics/flowredux/dsl/State.kt

```diff
@@ -0,0 +1,37 @@
+package com.freeletics.flowredux.dsl
+
+/**
+ * Allows to create [ChangeState] objects to change the state as a result of DSL
+ * methods like [InStateBuilderBlock.on] or  [InStateBuilderBlock.onEnter].
+ *
+ * The [snapshot] property can be used to access the current state at the time the action
+ * or event happened. Note that it might be outdated by the time the code runs and should
+ * never be used within [mutate] or to create an object that is passed to [override]
+ */
+public class State<InputState : Any>(
+    val snapshot: InputState
+) {
+    /**
+     * Use this function if you want to "mutate" the current state by copying the old state
+     * and modify some properties in the copy of the new state. A common use case is to call
+     * .copy() on your state defined as data class.
+     */
+    public fun <S : Any> mutate(reducer: InputState.() -> S): ChangeState<S> {
```

Comment:
Do we want to introduce 2 different methods? one `mutate(reducer : InputState.() -> InputState)` and one `override(InputState.() -> S)` and a shortcut `override(newState : S)`. I think we also discussed in this context to maybe rename `override` to  `transition`. 

I have not strong opinion, just want to ask if you have considered it and explicitly would like to stick with what is part of this PR.

Reply:
Having two methods was a solution for not having `InputState` anywhere but `State` already solves that. With that solved we don't need an extra method with a more concrete return type as the return type can be inferred by the compiler anyways, so you should never need to specify `S` specifically. The compiler will also guarantee that `S` is a super type of `InputState` when you use `mutate` or a method returning the result of `mutate` inside the DSL even though we don't have the `InputState : S` hierarchy encoded directly in `State`. 

So if we introduce an extra method it would simply be because we think it looks better when reading it and I don't have a strong opinion on that (same for whether we should call it override or transition).

Reply:
Correct, it's purely about readability (with mutate I expect a "change" or mutation of the current "state value" and not returning an entirely different state object. 

Moreover, if we go with `Snapshot` as nane I would be in favor of renaming override to transition and introduce the 2 options variations `transition(newState : S)` and `transition(reduce: InputState.() -> S)` and keep mutation only for mutating same type